### PR TITLE
Support spline-based ground modeling functionality

### PR DESCRIPTION
**Pull request by davidirobinson**
Add callback functions to draw the ground surface class image (containing free space, obstacles and out of bounds pixels corresponding to the disparity image) and draw a subsampled spline pointcloud in RVIZ.

Note that to switch the ground surface algorithm on (provided it's available and running on the camera), one must update the camera profile by adding `gen.const("ground_surface", int_t, 4, "")` to `multisense.cfg` and using `rqt_reconfigure`.

See further instructions for running here: https://github.com/carnegierobotics/stereo_ground_surface#visualizing-results

### CODE REVIEW

File: multisense_ros/src/camera.cpp [hunk 1]
```diff
@@ -212,6 +217,257 @@ cv::Vec3b u_interpolate_color(double u, double v, const cv::Mat &image)
     return result;
 }
 
+Eigen::Matrix<uint8_t, 3, 1> groundSurfaceClassToPixelColor(const uint8_t value)
```
Comment (mattalvarado) [inline multisense_ros/src/camera.cpp]: Maybe we should move some of the functions to a `spline.cc` or `groundsurface.cc` utility

File: multisense_ros/src/camera.cpp [hunk 2]
```diff
@@ -328,6 +340,15 @@ Camera::Camera(Channel* driver, const std::string& tf_prefix) :
     raw_cam_config_pub_ = calibration_nh_.advertise<multisense_ros::RawCamConfig>(RAW_CAM_CONFIG_TOPIC, 1, true);
     histogram_pub_      = device_nh_.advertise<multisense_ros::Histogram>(HISTOGRAM_TOPIC, 5);
 
+    // TODO(drobinson): guard this inside appropriate conditional
```
Comment (davidirobinson) [inline multisense_ros/src/camera.cpp]: I'm not sure whether I should advertise these streams by default.

Comment (mattalvarado) [reply multisense_ros/src/camera.cpp]: I would check the device info and at least make sure the camera is a S27/S30

File: multisense_ros/src/camera.cpp [hunk 3]
```diff
@@ -1873,6 +1904,97 @@ void Camera::colorizeCallback(const image::Header& header)
...
+        const uint32_t imageSize = 3 * height * width;
+
+        ground_surface_image_.data.resize(imageSize);
+
+        ground_surface_image_.header.frame_id = frame_id_rectified_left_;
+        ground_surface_image_.header.stamp    = t;
+        ground_surface_image_.height          = height;
+        ground_surface_image_.width           = width;
+
+        ground_surface_image_.encoding        = sensor_msgs::image_encodings::RGB8;
```
Comment (mattalvarado) [inline multisense_ros/src/camera.cpp:1946]: Did did not realize were were sending a raw color image over the wire. Does it make sense to send a 8bit mono image where each class has a its unique id. We can then convert that to a RGB image here.

I am approaching this from the angle of saving bandwidth on the wire.

Comment (mattalvarado) [reply multisense_ros/src/camera.cpp:1946]: Never mind. Should have read the next few lines

File: multisense_ros/src/ground_surface_utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    basisCoefficients(1) = 0.0;
+    basisCoefficients(2) = 0.0;
+    basisCoefficients(3) = 1.0 / 6.0;
+    ret(3) = basisCoefficients;
+
+    return ret;
+}
+
+template <class FloatT>
+FloatT getSplineValue(
```
Comment (mattalvarado) [inline multisense_ros/src/ground_surface_utilities.cpp:110]: You should move this template definition to the header file

File: multisense_ros/src/ground_surface_utilities.cpp [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    const auto minX = xzCellOrigin[0] + xzCellSize[0];
+    const auto minZ = xzCellOrigin[1] + xzCellSize[1];
+    const auto maxX = xzLimit[0];
+    const auto maxZ = xzLimit[1];
+
+    // Precompute number of points
+    size_t numPoints = 0;
+    for (float x = minX; x < maxX; x += drawResolution)
+        for (float y = minZ; y < maxZ; y += drawResolution)
+            numPoints++;
```
Comment (mattalvarado) [inline multisense_ros/src/ground_surface_utilities.cpp]: Can you do this without two for loops? Wouldn't something like this work?

`const size_t numPoints = std::floor((maxX - minX) / drawResolution) * std::floor((maxY - minY) / drawResolution)`

File: multisense_ros/src/ground_surface_utilities.cpp [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+    std::vector<Eigen::Vector3f> points;
```
Comment (mattalvarado) [inline multisense_ros/src/ground_surface_utilities.cpp]: You should be good with a vector of Vector3f points, but this is probably worth a read: https://eigen.tuxfamily.org/dox/group__TopicStlContainers.html

**Review by mattalvarado — COMMENTED** [review]
Overall this looks great! Thanks @davidirobinson! I made a few minor comments, but hopefully the are easy to fix

File: multisense_ros/src/ground_surface_utilities.cpp (hunk 4 above)
Comment (davidirobinson) [reply multisense_ros/src/ground_surface_utilities.cpp:110]: Just checking in about this - `getSplineValue` is only used inside `ground_surface_utilities.cpp` and is in an anonymous namespace. Basically I was trying to abstract it away and only expose the top-level functions that are required to draw the spline (see `multisense/multisense_ros/include/multisense_ros/ground_surface_utilities.h`).

But, if you think it's still best to move to the header file I can totally do that.

File: multisense_ros/src/ground_surface_utilities.cpp (hunk 5 above)
Comment (davidirobinson) [reply multisense_ros/src/ground_surface_utilities.cpp]: Absolutely, thanks for picking up on this (lazy programming on my behalf)!

File: multisense_ros/src/ground_surface_utilities.cpp (hunk 6 above)
Comment (davidirobinson) [reply multisense_ros/src/ground_surface_utilities.cpp]: Good catch on this, best to be safe!

File: multisense_ros/src/ground_surface_utilities.cpp (hunk 4 above)
Comment (mattalvarado) [reply multisense_ros/src/ground_surface_utilities.cpp:110]: Gotcha. That should be fine then.

**Review by mattalvarado — DISMISSED** [review]
This looks great. Thanks for integrating this!

**Review by mattalvarado — APPROVED** [review]
Frustum check looks good!

**Pull request merged**